contracts/Joint.sol

```diff
@@ -159,85 +185,302 @@ contract Joint {
+
+            (address sellToken, uint256 sellAmount) =
+                calculateSellToBalance(
+                    currentA,
+                    currentB,
+                    investedA,
+                    investedB
+                );
+
+            if (sellToken != address(0) && sellAmount != 0) {
+                uint256 buyAmount =
+                    sellCapital(
+                        sellToken,
+                        sellToken == tokenA ? tokenB : tokenA,
+                        sellAmount
+                    );
+                emit SellToBalance(sellToken, sellAmount, buyAmount);
+
+                if (sellToken == tokenA) {
+                    currentA = currentA.sub(sellAmount);
+                    currentB = currentB.add(buyAmount);
+                } else {
+                    currentB = currentB.sub(sellAmount);
+                    currentA = currentA.add(buyAmount);
+                }
+
+                (ratioA, ratioB) = getRatios(
+                    currentA,
+                    currentB,
+                    investedA,
+                    investedB
+                );
+                emit Ratios(ratioA, ratioB, "after balance");
+            }
+        }
 
-    function harvest() external onlyGuardians {
-        // IF tokenA or tokenB are rewards, we would be swapping all of it
-        // Let's save the previous balance before claiming
-        uint256 previousBalanceOfReward = balanceOfReward();
+        investedA = investedB = 0;
 
-        // Gets the reward from the masterchef contract
-        getReward();
-        uint256 rewardProfit = balanceOfReward().sub(previousBalanceOfReward);
-        if (rewardProfit > 0) {
-            swapReward(rewardProfit);
+        if (returnFunds) {
+            _returnLooseToProviders();
         }
+    }
 
+    function adjustPosition(bool reinvest) external onlyProviders {
         // No capital, nothing to do
-        if (balanceOfA() == 0 && balanceOfB() == 0) {
+        if (balanceOfA() == 0 || balanceOfB() == 0) {
             return;
         }
 
         if (reinvest) {
-            createLP();
+            require(
+                balanceOfStake() == 0 &&
+                    balanceOfPair() == 0 &&
+                    investedA == 0 &&
+                    investedB == 0
+            ); // don't create LP if we are already invested
+
+            (investedA, investedB, ) = createLP();
             depositLP();
+        }
+    }
+
+    function estimatedTotalAssetsAfterBalance()
+        public
+        view
+        returns (uint256 _aBalance, uint256 _bBalance)
+    {
+        uint256 rewardsPending = pendingReward();
+
+        (_aBalance, _bBalance) = balanceOfTokensInLP();
+
+        if (reward == tokenA) {
+            _aBalance = _aBalance.add(rewardsPending);
+        } else if (reward == tokenB) {
+            _bBalance = _bBalance.add(rewardsPending);
+        } else if (rewardsPending != 0) {
+            address swapTo = findSwapTo(reward);
```

Comment:
not sure it makes sense in terms of gas. 

you will almost always need to rebalance in one direction, it may be better to sell rewards for the token you need to buy during rebalance

Reply:
I always try to sell the reward to WETH as it's the intermediate no matter what. I should research if the cost of two swaps `reward => weth => token` is much greater than the same path in one swap.